salt/modules/archive.py

```diff
@@ -413,12 +413,20 @@ def _unsupported_format(archive_format):
             ret = {'dirs': sorted(dirs),
                    'files': sorted(files),
                    'links': sorted(links)}
-            ret['top_level_dirs'] = [x for x in ret['dirs']
-                                     if x.count('/') == 1]
-            ret['top_level_files'] = [x for x in ret['files']
-                                      if x.count('/') == 0]
-            ret['top_level_links'] = [x for x in ret['links']
-                                      if x.count('/') == 0]
+            if six.PY2:
+                ret['top_level_dirs'] = [x for x in ret['dirs']
+                                         if x.decode('utf-8').count('/') == 1]
+                ret['top_level_files'] = [x for x in ret['files']
+                                          if x.decode('utf-8').count('/') == 0]
+                ret['top_level_links'] = [x for x in ret['links']
+                                          if x.decode('utf-8').count('/') == 0]
+            else:
+                ret['top_level_dirs'] = [x for x in ret['dirs']
+                                         if x.count('/') == 1]
+                ret['top_level_files'] = [x for x in ret['files']
+                                          if x.count('/') == 0]
+                ret['top_level_links'] = [x for x in ret['links']
+                                          if x.count('/') == 0]
```

Comment:
It looks like the `getmembers` method within `tarfile` is returning paths as `str` types on Python 2. If this is the case, then the specific handling for Python 2 and 3 won't be necessary if you use `salt.utils.data.decode_list()` on each list when initializing `ret`:

```python
ret = {'dirs': sorted(salt.utils.data.decode_list(dirs)),
       'files': sorted(salt.utils.data.decode_list(files)),
       'links': sorted(salt.utils.data.decode_list(links))}
```

This will ensure that all paths are unicode types, and the list comprehensions which generate the `top_level_*` keys will therefore not cause a `UnicodeDecodeError`.

`salt.utils.data.decode()` would also work here, but it is a general-purpose function that does type checking and would end up simply calling `salt.utils.data.decode_list()` under-the-hood. Since we know that `dirs`, `files`, and `links` are all lists, using `salt.utils.data.decode_list()` would be marginally faster here.